### ISSUE TITLE
Support of Boost 1.59 test decorators as additional traits ?

### ISSUE THREAD

**Issue by maschwie** [etas/vs-boost-unit-test-adapter#64]
Hi,
we like to use the new “decorators” to organize our boost test suite in the same way as it can be done with NUnit category attributes.
http://www.boost.org/doc/libs/1_59_0/libs/test/doc/html/boost_test/tests_organization/decorators.html

For that purpose it is necessary that the adapter transfers all "decorations" additionally to the "test suite" as  "visual studio test explorer trait" to visual studio (like NUnit: it transfers every category as trait). Thus a boost test method like "BOOST_AUTO_TEST_CASE(test_case2 ..." may be assigned to several traits.

For the example below  from the boost documentation we would need >= 3 traits in the visual studio test explorer:
Trait1: utf_label_trivial
Trait2: utf_label_cmp
Trait3: TestSuite_ExampleSuite
(Trait3 is already created by the Boost 1.55 Version of the “Boost Unit Test Adapter”)

Boost Docu Example
http://www.boost.org/doc/libs/1_59_0/libs/test/doc/html/boost_test/tests_organization/decorators.html
BOOST_AUTO_TEST_SUITE(ExampleSuite)
BOOST_AUTO_TEST_CASE(test_case2,
  * utf::label("trivial")
  * utf::label("cmp")
  * utf::description("testing equality of ones"))
{

You can also contact me via "martin.schwienbacher@etas.com"

Mit freundlichen Grüßen / Best regards,
Martin Schwienbacher

ETAS GmbH
Engineering Software (ETAS/ESW1)
Borsigstrasse 14
70469 Stuttgart
Germany
www.etas.com

Phone +49(711)3423-2636
martin.schwienbacher@etas.com

Managing Directors: Friedhelm Pickhard, Bernd Hergert, Christopher White
Chairman of the Supervisory Board: Dr. Walter Schirm
Registered Office: Stuttgart; Registration Court: Amtsgericht Stuttgart HRB 19033

**Comment by guwirth** [comment]
Found a ms page describing it: http://blogs.msdn.com/b/visualstudioalm/archive/2012/11/09/how-to-manage-unit-tests-in-visual-studio-2012-update-1-part-1-using-traits-in-the-unit-test-explorer.aspx

**Comment by netspiri** [comment]
It is possible to add this feature but this will require a change in design for the list content feature (#24). The '--list_content' output does not provide any of the mentioned information so we need to rely on source discovery.

If the new default behaviour is to make use of '--list_content', users will need a way to explicitly turn off this discovery method, or select one based on their needs. This will require extra effort from a user's point of view. The approach we use for this kind of requirement would be to add additional options to the configuration file.

Is this an acceptable compromise?

**Comment by guwirth** [comment]
Is it another option to use own test runner or overwrite **--list_content** somehow to get the needed additional information? Think on the long run it's not a good idea to maintain two different approaches.

**Comment by maschwie** [comment]
> configuration file.
Maybe it is best to offer the feature in "configurable" way.
For us it would be ok if the new traits for "decorator labels" have to be enabled explicitly. 
I also think it's important that the visual studio user still can choose the legacy behavior (only one trait per test case that is derived from the test suite) - the"traits" are also important for Team Foundation Server Build Automation (see Article linked by guwirth) and legacy scripts may fail when the default number of traits per test case is changed.

**Comment by guwirth** [comment]
Here is a description from boost test:
http://www.boost.org/doc/libs/1_59_0/libs/test/doc/html/boost_test/runtime_config/test_unit_filtering.html#ref_command_line_control
Question is how to get a list of **utf::label** to map them to the **traits**?

**Comment by raffienficiaud** [comment]
Like this
http://www.boost.org/doc/libs/master/libs/test/doc/html/boost_test/utf_reference/rt_param_reference/list_labels.html

**Comment by netspiri** [comment]
@guwirth Mapping labels as traits is not an issue. The procedure is similar to how we currently treat test suites so it is just a matter of acquiring the labels in the first place.

@raffienficiaud Hello and thank you for your input.

The problem I see with `--list_labels` is that there is no connection between the test cases and the labels themselves. However, it seems that this connection is available in `--list_content=DOT` which should suffice for our needs.

Question: Is `--list_content=DOT` available for Boost 1.59 i.e. current Boost Test 3? Documentation is not available for it but it seems that the function does work. Is it recommended to use this feature?

**Comment by raffienficiaud** [comment]
Right, maybe we should add it.

`--list_content=DOT` is available for boost 1.59 (I did not have the time to document it), some minor bug fixes occurred for boost 1.60.

**Comment by netspiri** [comment]
@raffienficiaud Thank you for the prompt response.

We will consider `--list_content=DOT` since it is much more appropriate for our needs than `--list_content(=HRF)` which we are currently using.

**Comment by maschwie** [comment]
Thank's for the quick responses.

@guwirth
the possibility to list / filter test suites for test cases according to "decorations" is also very useful for "quality reporting" about the content of large test suites. 

For NUnit (McdCore) i have created custom category attributes. Via reflection it will be quite easy to extract all test methods and namespaces from an assembly that are attributed as "systemtest" and also add the scrum user story to the report.

For Boost i see a similar possibility:
If we are able to list e.g. all Boost Tests that are decorated as label "systemtest" or "regressiontest"  and also dump all other available decorations into a csv file, such reports can also be extracted for Boost Suites.  (we can add the user story number or the BugID also as label or another type of decorator)

**Comment by guwirth** [comment]
As an user of BUTA I want to be able to assign one or more labels to my unit test(s) to be able to search for or sort it afterwards in VS with the traits mechanism. 

Hints:
* support `--list_content=DOT`
* map utf::label to traits
* boost test 1.59+

**Comment by guwirth** [comment]
This issue is the most important one from our side. Do you have already an idea when it's done?

**Comment by maschwie** [comment]
I totally agree :-) 

btw. good news from the non-ui command line tooling and the jenkins servers. Since "Morning Friday 7. Jan 2016" wolfgang has really switched the whole productive test suite to Boost Decorators (in parallel to .NET Attributes) including the gerrit check-in gates ... so far all works fine and all linux and windows servers only execute the test case types they should.

**Comment by maschwie** [comment]
I have to extend this requirement, else it will not be usable for a non-small Test Suite (> 10 C ++ Test Projects)
See also item: "Configurable Test Runner Process Spawning" #81

Extension:
When the trait filter is used to select all test cases of a trait or a subset, the boost unit test adapter shall execute the test cases matching the filters by the minimal number of "*.Boostd.exe" test runner process spawns.

Example: when the trait "component test" results in 50 test cases (having the decorator "component test") from 21 suites from 10 different VC C++ Test Projects of the actual solution.

When executing the trait tests shall only cause: 10 Test Runner Process "spawns"
(not 21 and definitely not 50 what is actually the default when using the vs test explorer)

**Comment by guwirth** [comment]
Still issue with highest priority. Any idea if this will be solved soon?

**Comment by maschwie** [comment]
As our test suite is growing in #test cases it would be really nice to have a first implementation.
Please treat this issue with higher prio.

**Comment by guwirth** [comment]
@maschwie please do some testing with #109

**Comment by maschwie** [comment]
Tested #109 Feature: "Via the DOT representation, labels are now available as traits (refer to attached screenshot)" 

Using VS 2015 with our Suite and approx 1200 decorated boost test cases (plus a few hundred NUnit cases in the same solution)
- Visualization in VS Test Explorer works fine. :-)
- Filtering via VS Test Explorer trait filter expressions works fine. :-)
:-)
